components/base/sensorcontrolled/spin.go

```diff
@@ -105,162 +52,88 @@ func (sb *sensorBase) stopSpinWithSensor(
-		defer ticker.Stop()
+	for {
+		// check if we want to poll the sensor at all
+		// other API calls set this to false so that this for loop stops
 
-		for {
-			// check if we want to poll the sensor at all
-			// other API calls set this to false so that this for loop stops
-			if !sb.isPolling() {
-				ticker.Stop()
-			}
-
-			if err := ctx.Err(); err != nil {
-				ticker.Stop()
-				return
-			}
+		if !sb.isPolling() {
+			ticker.Stop()
+			sb.logger.CWarn(ctx, "Spin call interrupted by another running api")
+			return nil
+		}
 
-			select {
-			case <-ctx.Done():
-				ticker.Stop()
-				return
-			case <-ticker.C:
-				// imu readings are limited from 0 -> 360
-				currYaw, err := getCurrentYaw(sb.orientation)
-				if err != nil {
-					errCounter++
-					if errCounter > 100 {
-						sb.logger.CError(ctx, errors.Wrap(
-							err,
-							"imu sensor unreachable, 100 error counts when trying to read yaw angle, stopping base"))
-						if err = sb.Stop(ctx, nil); err != nil {
-							return
-						}
-					}
-					continue
-				}
-				errCounter = 0 // reset reading error count to zero if we are successfully reading again
+		if err := ctx.Err(); err != nil {
+			ticker.Stop()
+			return err
+		}
 
-				atTarget, overShot, minTravel := getTurnState(currYaw, startYaw, targetYaw, dir, angleDeg, errBound)
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-ticker.C:
+			orientation, err := sb.orientation.Orientation(ctx, nil)
+			if err != nil {
+				return err
+			}
 
-				if sensorDebug {
-					sb.logger.CDebugf(ctx, "minTravel %t, atTarget %t, overshot %t", minTravel, atTarget, overShot)
-					sb.logger.CDebugf(ctx, "angleDeg %.2f", angleDeg)
-					sb.logger.CDebugf(ctx,
-						"currYaw %.2f, startYaw %.2f, targetYaw %.2f",
-						currYaw, startYaw, targetYaw)
-				}
+			currYaw := rdkutils.RadToDeg(orientation.EulerAngles().Yaw)
+			angErr, prevMovedAng = getAngError(currYaw, prevAngle, prevMovedAng, angleDeg)
+			prevAngle = currYaw
 
-				// poll the sensor for the current error in angle
-				// check if we've overshot our target by the errTarget value
-				// check if we've travelled at all
-				if minTravel && (atTarget || overShot) {
-					if sensorDebug {
-						sb.logger.CDebugf(ctx,
-							"stopping base with errAngle:%.2f, overshot? %t",
-							math.Abs(targetYaw-currYaw), overShot)
-					}
+			if math.Abs(angErr) < boundCheckTarget {
+				return sb.Stop(ctx, nil)
+			}
+			angVel := calcAngVel(angErr, degsPerSec)
 
-					if err := sb.Stop(ctx, nil); err != nil {
-						return
-					}
-				}
+			if err := sb.updateControlConfig(ctx, 0, angVel); err != nil {
+				return err
+			}
 
-				if time.Since(startTime) > timeOut {
-					sb.logger.CWarn(ctx, "exceeded time for Spin call, stopping base")
-					if err := sb.Stop(ctx, nil); err != nil {
-						return
-					}
-					return
+			if time.Since(startTime) > timeOut {
+				sb.logger.CWarn(ctx, "exceeded time for Spin call, stopping base")
```

Comment:
have you seen situations in which we need the timer because the base is spinning longer than it should? I don't think motor relies on a timer for it's movement calls, so I wonder how necessary it is to have here

Reply:
the base can spin for longer than it should if the sensors are reporting the incorrect positions or if the motors are not actually setting power(like the pwm pin on the driver gets unplugged). it generally shouldn't be an issue under normal operation tho